src/syscall/libc_wasip2.go

```diff
@@ -689,133 +533,89 @@ func init() {
-)
-
-type __wasi_filesystem_descriptor_flags uint32
-
-const (
-	__wasi_filesystem_descriptor_flag_read __wasi_filesystem_descriptor_flags = 1 << iota
-	__wasi_filesystem_descriptor_flag_write
-	__wasi_filesystem_descriptor_flag_file_integrity_sync
-	__wasi_filesystem_descriptor_flag_data_integrity_sync
-	__wasi_filesystem_descriptor_flag_requested_write_sync
-	__wasi_filesystem_descriptor_flag_mutate_directory
-)
-
-type __wasi_filesystem_open_at_flags uint32
-
-const (
-	__wasi_filesystem_open_at_flags_create __wasi_filesystem_open_at_flags = 1 << iota
-	__wasi_filesystem_open_at_flags_directory
-	__wasi_filesystem_open_at_flags_exclusive
-	__wasi_filesystem_open_at_flags_truncate
-)
-
-func findPreopenForPath(path string) (__wasi_filesystem_descriptor, string) {
-	for k, v := range __wasi_filesystem_preopens {
+// FIXME(ydnar): opening a stripped path fails, so ignore it.
+func findPreopenForPath(path string) (types.Descriptor, string) {
+	for k, v := range wasiPreopens {
 		if strings.HasPrefix(path, k) {
 			path = strings.TrimPrefix(path, k+"/")
 			return v, path
 		}
 	}
-	return __wasi_cwd_descriptor, path
+	return wasiCWD, path
 }
 
 // int open(const char *pathname, int flags, mode_t mode);
 //
 //go:export open
 func open(pathname *byte, flags int32, mode uint32) int32 {
-	pathLen := strlen(unsafe.Pointer(pathname))
-	pathStr := _string{
-		pathname, uintptr(pathLen),
-	}
-	path := *(*string)(unsafe.Pointer(&pathStr))
-
-	dir, path := findPreopenForPath(path)
+	path := goString(pathname)
+	dir, _ := findPreopenForPath(path)
 
-	var dflags __wasi_filesystem_descriptor_flags
+	var dflags types.DescriptorFlags
 	if (flags & O_RDONLY) == O_RDONLY {
-		dflags |= __wasi_filesystem_descriptor_flag_read
+		dflags |= types.DescriptorFlagsRead
 	}
 	if (flags & O_WRONLY) == O_WRONLY {
-		dflags |= __wasi_filesystem_descriptor_flag_write
+		dflags |= types.DescriptorFlagsWrite
 	}
 
-	var oflags __wasi_filesystem_open_at_flags
+	var oflags types.OpenFlags
 	if flags&O_CREAT == O_CREAT {
-		oflags |= __wasi_filesystem_open_at_flags_create
+		oflags |= types.OpenFlagsCreate
 	}
 	if flags&O_DIRECTORY == O_DIRECTORY {
-		oflags |= __wasi_filesystem_open_at_flags_directory
+		oflags |= types.OpenFlagsDirectory
 	}
 	if flags&O_EXCL == O_EXCL {
-		oflags |= __wasi_filesystem_open_at_flags_exclusive
+		oflags |= types.OpenFlagsExclusive
 	}
 	if flags&O_TRUNC == O_TRUNC {
-		oflags |= __wasi_filesystem_open_at_flags_truncate
+		oflags |= types.OpenFlagsTruncate
 	}
 
 	// By default, follow symlinks for open() unless O_NOFOLLOW was passed
-	var pflags __wasi_filesystem_path_flags = __wasi_filesystem_path_flag_symlink_follow
-	if flags&O_NOFOLLOW == 0 {
-		pflags &^= __wasi_filesystem_path_flag_symlink_follow
-	}
-
-	pathWasiStr := __wasi_string{
-		unsafe.Pointer(pathname), uint32(pathLen),
+	var pflags types.PathFlags = types.PathFlagsSymlinkFollow
+	if flags&O_NOFOLLOW == 0 { // FIXME(ydnar): is == 0 right?
```

Comment:
Oops, you're right.  If the `O_NOFOLLOW` was passed then this should be `== O_NOFOLLOW`.